Split mathematical program test

hongkai-dai opened the pull request:
The purpose of this PR is to split the tests inside `mathematical_program_test.cc`. Previously it contains tests on adding constraints/variables, that do not depend on solvers; it also contains tests that require calling solver solver to solve the problem. Moreover, we support different approaches in adding constraints , with or without the symbolic form. As a result, we end up with a lot of code duplication, to formulate the same optimization problem by adding constraints in different ways.

This PR has more than 500 lines, but a lot of it is just to split the original file into different smaller files.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4866)
<!-- Reviewable:end -->

hongkai-dai commented:
+@soonho-tri for feature review.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-:-KbC4S0t9q04jpHvaQcE:bboswjz)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 12 files at r1.
Review status: 1 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1339 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbC60sGGpF3SHImQASL:-KbC60sGGpF3SHImQASM:b-xkfnlh) ([raw file](https://github.com/robotlocomotion/drake/blob/57d97b6e89815d3f99c679b25ad1dc66ca6d3ae1/drake/solvers/mathematical_program.h#L1339)):*
> ```objc
>    * @tparam Derived An Eigen::Matrix with symbolic::Variable as the scalar
>    * type.
>    * Derived::ColsAtCompileTime == 1.
> ```

I think it's better to merge this line into the previous line?

---

*[drake/solvers/mathematical_program.h, line 1365 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbC6dKKqGy9frjyqFC_:-KbC6dKKqGy9frjyqFCa:b-xkfnlh) ([raw file](https://github.com/robotlocomotion/drake/blob/57d97b6e89815d3f99c679b25ad1dc66ca6d3ae1/drake/solvers/mathematical_program.h#L1365)):*
> ```objc
>    * @tparam Derived An Eigen::Matrix with symbolic::Variable as the scalar
>    * type.
>    * Derived::ColsAtCompileTime != 1.
> ```

I think it's better to merge this line into the previous line?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
The new code in this PR is
1. The block in `mathematical_program.h`  `MathematicalProgram::AddBoundingBoxConstraint`. The previous code cannot handle when the input variable is a block of a matrix. The new code can.
2. The tests in `optimization_example.h` and `optimization_example.cc` are taken from `mathematical_program_test.cc`. Now each test example generates a `MathematicalProgram` object, with or without using symbolic constraints. These test examples are solved by nonlinear solvers (SNOPT, NLopt, IPOPT), but some of the examples can also be solved by other solvers (like in `linear_system_solver_test.cc`).

---

Review status: 1 of 12 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-:-KbC9HNCs3121xl6Vdzs:b3d6rtq)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1339 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbC60sGGpF3SHImQASL:-KbCAmDNRYX9B76ZQBS0:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/57d97b6e89815d3f99c679b25ad1dc66ca6d3ae1/drake/solvers/mathematical_program.h#L1339)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I think it's better to merge this line into the previous line?
</blockquote></details>

Done. Thanks!

---

*[drake/solvers/mathematical_program.h, line 1365 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbC6dKKqGy9frjyqFC_:-KbCAnePGsHYtltqGkrG:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/57d97b6e89815d3f99c679b25ad1dc66ca6d3ae1/drake/solvers/mathematical_program.h#L1365)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I think it's better to merge this line into the previous line?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 12 files at r1, 1 of 2 files at r2.
Review status: 3 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/solvers/test/CMakeLists.txt, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCBzEmxiVFQbmWLUwo:-KbCBzEmxiVFQbmWLUwp:bruz6xa) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/CMakeLists.txt#L25)):*
> ```CMake
> target_link_libraries(constraint_test drakeOptimization drakeCommon)
> 
> add_executable(nonlinear_program_test
> ```

I wish there is an easier way to write this but I don't think it's non-trivial because we need to add two `cc` files into one test, right? (this is a non-blocking comment)

---

*[drake/solvers/test/optimization_examples.h, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCA1eRfw40Tdn-C_iv:-KbCA1eRfw40Tdn-C_iw:bbduquf) ([raw file](https://github.com/robotlocomotion/drake/blob/57d97b6e89815d3f99c679b25ad1dc66ca6d3ae1/drake/solvers/test/optimization_examples.h#L6)):*
> ```objc
> #include <memory>
> 
> #include "drake/common/eigen_matrix_compare.h"
> ```

This can be moved to `optimization_examples.cc` file.

---

*[drake/solvers/test/optimization_examples.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCAT4aVhlwlVSHgfOm:-KbCAT4aVhlwlVSHgfOn:b493r3s) ([raw file](https://github.com/robotlocomotion/drake/blob/57d97b6e89815d3f99c679b25ad1dc66ca6d3ae1/drake/solvers/test/optimization_examples.h#L17)):*
> ```objc
> class LinearSystemExample1 {
>  public:
>   LinearSystemExample1();
> ```

META : add `= default`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCCSv4yqQG9ZyM86lJ:-KbCCSv56v3qdcnMf73P:b86aij3) (no related file):*
FWIW, this is tested on a Mac (using both of cmake and bazel).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
*[drake/solvers/test/optimization_examples.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCAT4aVhlwlVSHgfOm:-KbCFBJ9U-di0SW4KM1T:b-tatwal) ([raw file](https://github.com/robotlocomotion/drake/blob/57d97b6e89815d3f99c679b25ad1dc66ca6d3ae1/drake/solvers/test/optimization_examples.h#L17)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

META : add `= default`?
</blockquote></details>

Never mind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 7 of 12 files at r1, 1 of 2 files at r2.
Review status: 11 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/solvers/test/mathematical_program_test_util.h, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCFTxkSor-ufrWLTZZ:-KbCFTxkSor-ufrWLTZ_:btgaii) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/mathematical_program_test_util.h#L19)):*
> ```objc
> namespace test {
> // TODO(#2274) Fix NOLINTNEXTLINE(runtime/references).
> void CheckSolverType(MathematicalProgram& prog,
> ```

Based on what it's doing in the body, it seems that this method can be better named as `CheckSolverName`, what do you think? 

---

*[drake/solvers/test/optimization_examples.h, line 90 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCCzGoBZ9w-D8y7Pao:-KbCCzGpRsvqx35oez6W:bwqphbm) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/optimization_examples.h#L90)):*
> ```objc
>   };
> 
>  public:
> ```

META nitpick: redundant `public:`

---

*[drake/solvers/test/optimization_examples.h, line 262 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCDzXoCYt9O7xdfUqp:-KbCDzXoCYt9O7xdfUqq:b-tkgr8a) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/optimization_examples.h#L262)):*
> ```objc
>   void AddNonSymbolicConstraint();
> 
>  private:
> ```

META nitpick: redundant `private:`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 12 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/solvers/test/nonlinear_program_test.cc, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCJwqvCoNqRGtVsfB1:-KbCJwqvCoNqRGtVsfB2:b-k61o8m) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/nonlinear_program_test.cc#L1)):*
> ```cpp
> #include <typeinfo>
> ```

I don't think `typeinfo` is needed in the file. Instead, please consider adding the following STL headers:

```c++
#include <array>       // std::array
#include <functional>  // std::function
#include <limits>      // std::numeric_limits
#include <map>         // std::map
#include <memory>      // std::shared_ptr
#include <stdexcept>   // std::runtime_error
#include <utility>     // std::pair
#include <vector>      // std::vector
```

---

*[drake/solvers/test/nonlinear_program_test.cc, line 9 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCJn0NuoWioc_LEpgo:-KbCJn0NuoWioc_LEpgp:bxb8bby) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/nonlinear_program_test.cc#L9)):*
> #include "drake/common/symbolic_expression.h"
> #include "drake/common/symbolic_variable.h"

no need.

---

*[drake/solvers/test/nonlinear_program_test.cc, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCJZItfDx1LkdeAtlm:-KbCJZItfDx1LkdeAtln:b1jk1vu) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/nonlinear_program_test.cc#L20)):*
> using Eigen::Dynamic;
> using Eigen::Ref;

no need.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (asides comments)

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-:-KbCLdS7OF-V3ixHXbIc:b-5lm0rb)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 6 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCCSv4yqQG9ZyM86lJ:-KbCFApYdQfhKbK0x8Ym:b-g2pwne) (no related file):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FWIW, this is tested on a Mac (using both of cmake and bazel).
</blockquote></details>

Thanks for testing it! +@sherm1 for platforem review. @sherm1, I summarized the new code in the comments above. 

---

*[drake/solvers/test/mathematical_program_test_util.h, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCFTxkSor-ufrWLTZZ:-KbCLsEsAf8-yDOA8Dny:b35wj23) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/mathematical_program_test_util.h#L19)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Based on what it's doing in the body, it seems that this method can be better named as `CheckSolverName`, what do you think? 
</blockquote></details>

Done. Good point!

---

*[drake/solvers/test/nonlinear_program_test.cc, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCJwqvCoNqRGtVsfB1:-KbCM0nzyzSZOIjXft3k:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/nonlinear_program_test.cc#L1)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I don't think `typeinfo` is needed in the file. Instead, please consider adding the following STL headers:

```c++
#include <array>       // std::array
#include <functional>  // std::function
#include <limits>      // std::numeric_limits
#include <map>         // std::map
#include <memory>      // std::shared_ptr
#include <stdexcept>   // std::runtime_error
#include <utility>     // std::pair
#include <vector>      // std::vector
```
</blockquote></details>

Done.

---

*[drake/solvers/test/nonlinear_program_test.cc, line 9 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCJn0NuoWioc_LEpgo:-KbCM64-ka5yFXJ6R3v2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/nonlinear_program_test.cc#L9)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> #include "drake/common/symbolic_expression.h"
> #include "drake/common/symbolic_variable.h"

no need.
</blockquote></details>

Done.

---

*[drake/solvers/test/nonlinear_program_test.cc, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCJZItfDx1LkdeAtlm:-KbCM8ErNrhVTi4lF7ok:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/nonlinear_program_test.cc#L20)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> using Eigen::Dynamic;
> using Eigen::Ref;

no need.
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.h, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCA1eRfw40Tdn-C_iv:-KbCEkJTr_6kphXm5UXK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/57d97b6e89815d3f99c679b25ad1dc66ca6d3ae1/drake/solvers/test/optimization_examples.h#L6)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

This can be moved to `optimization_examples.cc` file.
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.h, line 90 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCCzGoBZ9w-D8y7Pao:-KbCMJUAC9ZEfmgXQNkq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/optimization_examples.h#L90)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

META nitpick: redundant `public:`
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.h, line 262 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbCDzXoCYt9O7xdfUqp:-KbCMJsMdTs9u8ZK7m9_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a675406c787c17c00aa6156f7c1e334c860deacd/drake/solvers/test/optimization_examples.h#L262)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

META nitpick: redundant `private:`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform pass done, PTAL.

---

Reviewed 4 of 12 files at r1, 2 of 2 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1338 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDuOJ1ShO2gKbGujWW:-KbDuOJ1ShO2gKbGujWX:bdivxnd) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/mathematical_program.h#L1338)):*
> ```objc
>    * Adds the same scalar lower and upper bound to every variable in @p vars.
>    * @tparam Derived An Eigen::Matrix with symbolic::Variable as the scalar
>    * type. Derived::ColsAtCompileTime == 1.
> ```

BTW, sentence fragment. How about "as the scalar type and with a single column", or maybe call this an "Eigen Vector type"?

---

*[drake/solvers/mathematical_program.h, line 1353 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDurCtssZb0IirhWwI:-KbDurCtssZb0IirhWwJ:bkvdyf8) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/mathematical_program.h#L1353)):*
> ```objc
>         Derived::RowsAtCompileTime != Eigen::Dynamic &&
>                 Derived::ColsAtCompileTime != Eigen::Dynamic
>             ? Derived::RowsAtCompileTime * Derived::ColsAtCompileTime
> ```

BTW, don't we already know here that ColsAtCompileTime==1? If so this could be much simpler.

---

*[drake/solvers/mathematical_program.h, line 1363 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDv2vmuOUUv9SBVm1Y:-KbDv2vnU1uCsDUlmrFa:b-nwdx2y) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/mathematical_program.h#L1363)):*
> ```objc
>    * Adds the same scalar lower and upper bound to every variable in @p vars.
>    * @tparam Derived An Eigen::Matrix with symbolic::Variable as the scalar
>    * type. Derived::ColsAtCompileTime != 1.
> ```

BTW, fix sentence, maybe "with more than one column or an unknown number of columns"?

---

*[drake/solvers/test/equality_constrained_qp_solver_test.cc, line 30 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbE-buNulLZvFRV9G8E:-KbE-buNulLZvFRV9G8F:bvd1rlk) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/equality_constrained_qp_solver_test.cc#L30)):*
> ```cpp
>   auto x = prog.NewContinuousVariables<2>();
>   MatrixXd Q(2, 2);
> // clang-format off
> ```

BTW, comment indenting odd in this block.

---

*[drake/solvers/test/mathematical_program_test_util.h, line 18 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDz7iV6PTCC5Ocw3K0:-KbDz7iV6PTCC5Ocw3K1:b-9ugu30) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/mathematical_program_test_util.h#L18)):*
> ```objc
> namespace solvers {
> namespace test {
> // TODO(#2274) Fix NOLINTNEXTLINE(runtime/references).
> ```

Why does this need a styleguide exception?

---

*[drake/solvers/test/nonlinear_program_test.cc, line 42 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDzKZcLl_Aj3Z4A7I-:-KbDzKZd6rW_g9QVe33E:b-9ugu30) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/nonlinear_program_test.cc#L42)):*
> ```cpp
> namespace test {
> // TODO(#2274) Fix NOLINTNEXTLINE(runtime/references).
> void RunNonlinearProgram(MathematicalProgram& prog,
> ```

Why does this need a styleguide exception?

---

*[drake/solvers/test/nonlinear_program_test.cc, line 206 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDziW_OMYBmeQUiUCE:-KbDziW_OMYBmeQUiUCF:b-sc6v20) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/nonlinear_program_test.cc#L206)):*
> ```cpp
> 
>   template <typename ScalarType>
>   // TODO(#2274) Fix NOLINTNEXTLINE(runtime/references).
> ```

Why styleguide exception?

---

*[drake/solvers/test/nonlinear_program_test.cc, line 259 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDzmY3hXghvLp4RVk5:-KbDzmY3hXghvLp4RVk6:b-9dm7v6) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/nonlinear_program_test.cc#L259)):*
> ```cpp
> // Test that linear polynomial constraints get turned into linear constraints.
> // TODO(hongkai.dai): move this example to optimization_program_examples, add
> // the constraint in the symbolic form.
> ```

BTW, is this TODO still the plan? (and elsewhere in this file)

---

*[drake/solvers/test/optimization_examples.cc, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDyO0LyLPy6dRIcQGP:-KbDyO0MNE2Lnrsqh1q3:bu6drsv) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.cc#L31)):*
> ```cpp
> bool LinearSystemExample1::CheckSolution() const {
>   auto x_sol = prog_->GetSolution(x_);
>   if (!CompareMatrices(x_sol, b_, 1E-10, MatrixCompareType::absolute)) {
> ```

Rather than repeat this `1e-10` throughout the file, give it a meaningful symbolic constant name at the top and use that. (Even just `tol` would be better.)

---

*[drake/solvers/test/optimization_examples.h, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDvn-7aOQgFbQRoe5A:-KbDvn-7aOQgFbQRoe5B:b-tt9t9c) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L67)):*
> ```objc
> 
> /// This test comes from Section 2.2 of "Handbook of Test Problems in
> /// Local and Global Optimization."
> ```

BTW, can you give a more complete reference? Should give at least copyright date to insulate this from new editions where Section 2.2 might be different.

---

*[drake/solvers/test/optimization_examples.h, line 181 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwIEWxssLeh14_J71:-KbDwIEWxssLeh14_J72:b-3b1zlh) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L181)):*
> ```objc
> 
> /// This test comes from Section 3.4 of "Handbook of Test Problems in
> /// Local and Global Optimization.
> ```

BTW, more complete ref.

---

*[drake/solvers/test/optimization_examples.h, line 266 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwRsr3uUyPCyA-tWl:-KbDwRsr3uUyPCyA-tWm:bme4xwm) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L266)):*
> ```objc
> ///
> /// Which is from section 3.5 in
> ///   Handbook of Test Problems in Local and Global Optimization
> ```

BTW, ref should be more complete and consistent.

---

*[drake/solvers/test/optimization_examples.h, line 321 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwhRi3wm4HAzkV1ZF:-KbDwhRi3wm4HAzkV1ZG:bm5o5f6) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L321)):*
> ```objc
> 
>    protected:
>     // for just these two types, implementing this locally is almost cleaner...
> ```

BTW, cap punc (here & above)

---

*[drake/solvers/test/optimization_examples.h, line 335 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwtGL15l5OmM6AbWm:-KbDwtGL15l5OmM6AbWn:b-w0y03u) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L335)):*
> ```objc
>     void EvalImpl(
>         const Eigen::Ref<const Eigen::Matrix<ScalarType, Eigen::Dynamic, 1>>& x,
>         // TODO(#2274) Fix NOLINTNEXTLINE(runtime/references).
> ```

BTW, why the style exception here? Couldn't y be a nonconst pointer. Ugly, I agree, but I'm not seeing why this should be an exception.

---

*[drake/solvers/test/optimization_examples.h, line 358 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDxES4if3-j0mGGSZv:-KbDxES4if3-j0mGGSZw:b-4jkt8e) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L358)):*
> ```objc
> };
> 
> // An SOCP with Lorentz cone and rotated Lorentz cone constraints.
> ```

There is a mix of doxygen/non-doxygen comments in this file for no apparent reason.

---

*[drake/solvers/test/optimization_examples.h, line 363 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDxlrgUbInatT579ES:-KbDxlrgUbInatT579ET:bwelrhx) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L363)):*
> ```objc
> // We can solve the following SOCP with Lorentz cone constraint
> // min  t
> //  s.t t >= sqrt(x'*x)
> ```

BTW, if you want to get away from this Matlab-specific notation, Unicode does provide a lovely transpose character you can copy and past from here: xᵀ.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-:-KbE05wr2_IBO5hCrq-r:b-fzvf1a)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@sherm1 I have addressed your comments, PTAL. 

---

Review status: 6 of 12 files reviewed at latest revision, 23 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1338 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDuOJ1ShO2gKbGujWW:-KbE4KOCtA8qzDvwsWcK:b-6ijwgi) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/mathematical_program.h#L1338)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, sentence fragment. How about "as the scalar type and with a single column", or maybe call this an "Eigen Vector type"?
</blockquote></details>

Done. Use "Eigen Vector type" to emphasize it has one column at compile time.

---

*[drake/solvers/mathematical_program.h, line 1353 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDurCtssZb0IirhWwI:-KbE4cRrv4ZDabL8Uz4d:b-376vyb) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/mathematical_program.h#L1353)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, don't we already know here that ColsAtCompileTime==1? If so this could be much simpler.
</blockquote></details>

Done. Good catch.

---

*[drake/solvers/mathematical_program.h, line 1363 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDv2vmuOUUv9SBVm1Y:-KbE4lWP1VAyhLGgGFvO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/mathematical_program.h#L1363)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, fix sentence, maybe "with more than one column or an unknown number of columns"?
</blockquote></details>

Done.

---

*[drake/solvers/test/equality_constrained_qp_solver_test.cc, line 30 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbE-buNulLZvFRV9G8E:-KbE51Dvk7O3Er4SjiI2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/equality_constrained_qp_solver_test.cc#L30)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, comment indenting odd in this block.
</blockquote></details>

Done.

---

*[drake/solvers/test/mathematical_program_test_util.h, line 18 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDz7iV6PTCC5Ocw3K0:-KbE5FnUs7uMBKYKyB0Y:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/mathematical_program_test_util.h#L18)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Why does this need a styleguide exception?
</blockquote></details>

Done.

---

*[drake/solvers/test/nonlinear_program_test.cc, line 42 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDzKZcLl_Aj3Z4A7I-:-KbE67KEbJsQuDaj_rsJ:biw6cq6) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/nonlinear_program_test.cc#L42)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Why does this need a styleguide exception?
</blockquote></details>

Done. I copied it from existing code, but have changed to use a `MathematicalProgram` pointer.

---

*[drake/solvers/test/nonlinear_program_test.cc, line 206 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDziW_OMYBmeQUiUCE:-KbE6G3lO35K-LaxSE9o:bkss6l1) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/nonlinear_program_test.cc#L206)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Why styleguide exception?
</blockquote></details>

It is how `drake::detail::Function` defines the function prototype, as in https://github.com/RobotLocomotion/drake/blob/master/drake/solvers/function.h#L44

---

*[drake/solvers/test/nonlinear_program_test.cc, line 259 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDzmY3hXghvLp4RVk5:-KbE6Y8mXmVndEA6ZeUg:b-6on25d) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/nonlinear_program_test.cc#L259)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, is this TODO still the plan? (and elsewhere in this file)
</blockquote></details>

Yes it is still the plan.

---

*[drake/solvers/test/optimization_examples.cc, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDyO0LyLPy6dRIcQGP:-KbE6zmSQEws4h7ovhdM:b-wzemh3) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.cc#L31)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Rather than repeat this `1e-10` throughout the file, give it a meaningful symbolic constant name at the top and use that. (Even just `tol` would be better.)
</blockquote></details>

Done. Used 'tol_`.

---

*[drake/solvers/test/optimization_examples.h, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDvn-7aOQgFbQRoe5A:-KbE72Js2Y-wn23tHu5P:b-nvgg2r) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L67)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, can you give a more complete reference? Should give at least copyright date to insulate this from new editions where Section 2.2 might be different.
</blockquote></details>

I do not have this book, I think Russ has it. @russtedrake is http://www.springer.com/us/book/9780792358015 the book you are using?

---

*[drake/solvers/test/optimization_examples.h, line 181 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwIEWxssLeh14_J71:-KbE7T9Ke5KyhmUQicNq:bs4q9op) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L181)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, more complete ref.
</blockquote></details>

Ditto.

---

*[drake/solvers/test/optimization_examples.h, line 266 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwRsr3uUyPCyA-tWl:-KbE7YBoNdKfAHNc8FXU:bs4q9op) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L266)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, ref should be more complete and consistent.
</blockquote></details>

Ditto.

---

*[drake/solvers/test/optimization_examples.h, line 321 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwhRi3wm4HAzkV1ZF:-KbE7oZowk7U27Xw7NBp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L321)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, cap punc (here & above)
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.h, line 335 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwtGL15l5OmM6AbWm:-KbE89GVS9SoV6S4yQ2Z:b-ii4yyv) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L335)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, why the style exception here? Couldn't y be a nonconst pointer. Ugly, I agree, but I'm not seeing why this should be an exception.
</blockquote></details>

Done. I don't know why there is an exception. I copied this code from previous code. I just changed it to a non-const pointer.

---

*[drake/solvers/test/optimization_examples.h, line 358 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDxES4if3-j0mGGSZv:-KbE9G7FBjZYe8n7naBM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L358)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

There is a mix of doxygen/non-doxygen comments in this file for no apparent reason.
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.h, line 363 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDxlrgUbInatT579ES:-KbE9GULTaZdP4ePMVcS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L363)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, if you want to get away from this Matlab-specific notation, Unicode does provide a lovely transpose character you can copy and past from here: xᵀ.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-:-KbED2nbX87jBlKpBbMv:b-a8hiex)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform: a few small things left, then <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1360 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbHG_ATkKnwrN-iZYV1:-KbHG_ATkKnwrN-iZYV2:b-injrzw) ([raw file](https://github.com/robotlocomotion/drake/blob/2c05f51ac4c842172e72ab2c8f7175cddc6d86c1/drake/solvers/mathematical_program.h#L1360)):*
> ```objc
>    * @tparam Derived An Eigen::Matrix with symbolic::Variable as the scalar
>    * type. The matrix has unknown number of columns at compile time, or has
>    * more than one columns.
> ```

BTW, "more than one column." (singular)

---

*[drake/solvers/test/optimization_examples.h, line 321 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwhRi3wm4HAzkV1ZF:-KbHHGBzexDFseMND4De:b5pa0hh) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L321)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

no change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-:-KbHHkEZQk5zQdbj-mMB:b-r8vvrd)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 10 of 12 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1360 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbHG_ATkKnwrN-iZYV1:-KbHKxY9mDIwgkD4k-NX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2c05f51ac4c842172e72ab2c8f7175cddc6d86c1/drake/solvers/mathematical_program.h#L1360)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, "more than one column." (singular)
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.h, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDvn-7aOQgFbQRoe5A:-KbHMBDjoGNSSbJ5Lb3V:b-aynh8r) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L67)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I do not have this book, I think Russ has it. @russtedrake is http://www.springer.com/us/book/9780792358015 the book you are using?
</blockquote></details>

Done, I put the copyright date and the ISBN

---

*[drake/solvers/test/optimization_examples.h, line 181 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwIEWxssLeh14_J71:-KbHMF9B1dQInidxvaB4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L181)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Ditto.
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.h, line 266 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwRsr3uUyPCyA-tWl:-KbHMGyrOwJeYKO-lj1l:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L266)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Ditto.
</blockquote></details>

Done.

---

*[drake/solvers/test/optimization_examples.h, line 321 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4866#-KbDwhRi3wm4HAzkV1ZF:-KbHKokyN4ztodvhnLlq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/61785efac2e9f5e320f9f347776a3274f3002b24/drake/solvers/test/optimization_examples.h#L321)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

no change?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 2 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4866)*
<!-- Sent from Reviewable.io -->
